### PR TITLE
Provide a default metrics implementation using a LoggingProvider

### PR DESCRIPTION
**Pull request by EdColeman**
- Adds the MeterRegistryFactory to `..core.spi.metrics` and deprecates the one in `..core.metrics`
 - Modified the GENERAL_MICROMETER_FACTORY to be a list of class names. This would allow multiple provides to be used. For example, logging and statsd. Primarly this may help track down metrics issues.
 - set the default for mini cluster to log metrics. This should increase the visibility of metrics during development, but may be cluttering the logs.

This replaces #4218

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * properties by enabling {@code Property.GENERAL_MICROMETER_ENABLED} and providing this class for
+ * the {@code Property.GENERAL_MICROMETER_FACTORY}
+ * <p>
+ * The metrics will appear in the normal service logs with a named logger of
+ * {@code org.apache.accumulo.METRICS} at the INFO level. The metrics output can be directed to a
+ * file using standard logging configuration properties.
+ */
+public class SimpleLoggingMeterRegistryFactory implements MeterRegistryFactory {
+
+  private static final Logger LOG =
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:42]: Is this used?

File: core/src/test/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+class SimpleLoggingMeterRegistryFactoryTest {
+
+  @Test
+  public void createTest() {
+    SimpleLoggingMeterRegistryFactory factory = new SimpleLoggingMeterRegistryFactory();
+    assertInstanceOf(MeterRegistry.class, factory.create());
+  }
+
+  @Test
+  public void x() {
```
Comment (dlmarion) [inline core/src/test/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactoryTest.java:37]: This doesn't test anything, maybe remove?

File: test/src/main/resources/log4j2-test.properties [hunk 3]
```diff
@@ -152,29 +152,6 @@ logger.38.level = debug
 logger.39.name = org.apache.accumulo.manager.Manager
 logger.39.level = trace
 
-property.metricsFilename = ./target/test-metrics
```
Comment (dlmarion) [inline test/src/main/resources/log4j2-test.properties:155]: I'm wondering if we should undo this change now, so that the metrics go to a specific file for the ITs instead of polluting the server logs.

Comment (EdColeman) [reply test/src/main/resources/log4j2-test.properties:155]: If we leave it, it needs reworked - that specific line you highlighted anyway - the log should be based on the service name (it's already a prop in the file) - hard coding it as test there helped when we tailed the logs so that it was fixed and well known, but maybe not the best example.

I'd prefer keeping it - or having it somewhere discover-able, because it is a pain to get it set up correctly when you are not configuring log4j regularly.

File: core/src/test/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactoryTest.java (hunk 2 above)
Comment (EdColeman) [reply core/src/test/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactoryTest.java]: removed in 608490bfd3

File: core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  // named logger that can be configured using standard logging properties.
+  private static final Logger METRICS = LoggerFactory.getLogger("org.apache.accumulo.METRICS");
+
+  private final Consumer<String> metricConsumer = METRICS::info;
+
+  // defines the metrics update period, default is 60 seconds.
+  private final LoggingRegistryConfig lconf = c -> {
+    if (c.equals("logging.step")) {
+      return "60s";
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:53]: Is there any way to configure this to change the frequency?

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 5]
```diff
@@ -300,8 +300,11 @@ public enum Property {
       "Enables metrics functionality using Micrometer.", "2.1.0"),
   GENERAL_MICROMETER_JVM_METRICS_ENABLED("general.micrometer.jvm.metrics.enabled", "false",
       PropertyType.BOOLEAN, "Enables JVM metrics functionality using Micrometer.", "2.1.0"),
-  GENERAL_MICROMETER_FACTORY("general.micrometer.factory", "", PropertyType.CLASSNAME,
-      "Name of class that implements MeterRegistryFactory.", "2.1.0"),
+  GENERAL_MICROMETER_FACTORY("general.micrometer.factory",
+      "org.apache.accumulo.core.spi.metrics.SimpleLoggingMeterRegistryFactory",
+      PropertyType.CLASSNAMELIST,
+      "A comma separated list of one or more class names that implement MeterRegistryFactory.",
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:306]:
```suggestion
      "A comma separated list of one or more class names that implement MeterRegistryFactory.  Prior to 2.1.3 the default value was blank.  In 2.1.3 the default value was changed.",
```

File: core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java (hunk 4 above)
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:53]: It is currently not a property - so no.  60s is actually the default, but specifically setting it here shows how to change it, if so inclined.  I'm not sure how far we want to go with supporting this as a default - users should really be using a production solution, so bare-bones impl for this may be okay.

File: core/src/main/java/org/apache/accumulo/core/metrics/MetricsUtil.java [hunk 6]
```diff
@@ -105,14 +108,16 @@ public DistributionStatisticConfig configure(Meter.Id id,
         }
       };
 
-      Class<? extends MeterRegistryFactory> clazz =
-          ClassLoaderUtil.loadClass(factoryClass, MeterRegistryFactory.class);
-      MeterRegistryFactory factory = clazz.getDeclaredConstructor().newInstance();
+      for (String factoryName : getTrimmedStrings(factoryClasses)) {
+        Class<? extends MeterRegistryFactory> clazz =
+            ClassLoaderUtil.loadClass(factoryName, MeterRegistryFactory.class);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metrics/MetricsUtil.java:113]: If someone had the old Factory interface configured it seems like this would blow up.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 5 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java:306]: It would be good to explain the change in expected interface class in the docs and what the support for the old interface class is.

File: core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java (hunk 4 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:53]: Being able to configure the class but not pass any config seems cumbersome to me.  If a user configure their own factory and wants to set a server to pass metrics too are they expected to hard code the server address?

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:53]: The  config problem is not something to worry about this in PR as its a per-existing pattern.  I am just wondering how someone would use actually use this feature and it it requires recompiling java code for every metrics config change.  If the end to end user experience could be improved we could consider that when introducing a new interface though.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 5 above)
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Addressed in d8551b56e4

File: core/src/main/java/org/apache/accumulo/core/metrics/MetricsUtil.java (hunk 6 above)
Comment (EdColeman) [reply core/src/main/java/org/apache/accumulo/core/metrics/MetricsUtil.java]: Addressed in d8551b56e4

**Review by dlmarion — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 7]
```diff
@@ -300,8 +300,13 @@ public enum Property {
...
   GENERAL_MICROMETER_JVM_METRICS_ENABLED("general.micrometer.jvm.metrics.enabled", "false",
       PropertyType.BOOLEAN, "Enables JVM metrics functionality using Micrometer.", "2.1.0"),
-  GENERAL_MICROMETER_FACTORY("general.micrometer.factory", "", PropertyType.CLASSNAME,
-      "Name of class that implements MeterRegistryFactory.", "2.1.0"),
+  GENERAL_MICROMETER_FACTORY("general.micrometer.factory",
+      "org.apache.accumulo.core.spi.metrics.SimpleLoggingMeterRegistryFactory",
+      PropertyType.CLASSNAMELIST,
+      "A comma separated list of one or more class names that implement MeterRegistryFactory. Prior to"
+          + " 2.1.3 this was a single value and the default was an empty string.  In 2.1.3 the default "
+          + " was changed and it now can accept multiple class names.",
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java:308]: Could also add something about the change in classes.  Not sure if I got the packages correct.

```
In 2.1.3 a new type was introduced o.a.a.core.spi.metrics.MeterRegistryFacory and the existing type o.a.a.core.metrics.MeterRegistryFactory was deprecated.  For now classes that extend either type are accepted as a value to this property.
```

**Review by keith-turner — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java (hunk 4 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:53]: The  TestStatsDRegistryFactory example splits config across two places.  If a plugin follows this approach (its hard to think of any other approach a plugin could follow) then metrics config would end up being split across accumulo.properties file and accumulo-env.sh files.  I will open a follow on issue about potentially improving this.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/spi/metrics/SimpleLoggingMeterRegistryFactory.java:53]: Opened #4262

**Comment by EdColeman** [comment]
I marked this as a blocker for 2.1.3.  The SPI likely should go in some form into 2.1.3.  Rather that merge this now, I working issue #4262 that could have additional improvements.  If 2.1.3 release is ready, then what exists can be merged now and changes handled later.

**Comment by EdColeman** [comment]
Replaced with #4459
